presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -806,6 +802,26 @@ public void finishInsert(ConnectorSession session, ConnectorInsertTableHandle in
             for (CompletableFuture<?> fileRenameFuture : fileRenameFutures) {
                 MoreFutures.getFutureValue(fileRenameFuture, PrestoException.class);
             }
+
+            // clean remaining partition directories
+            Optional<Path> writePathRoot = locationService.writePath(handle.getLocationHandle(), Optional.empty());
+            if (writePathRoot.isPresent() && !writePathRoot.get().equals(locationService.targetPathRoot(handle.getLocationHandle()))) {
+                for (PartitionUpdate partitionUpdate : partitionUpdates) {
+                    if (writePathRoot.isPresent() && !partitionUpdate.getTargetPath().equals(partitionUpdate.getWritePath())) {
+                        String writePath = partitionUpdate.getWritePath();
+                        checkState(writePath.startsWith(writePath), "Partition updated was write stored outside insert write path");
```

Comment:
Using `startWith` for `toString` of Path is generally not a good idea. I would write a helper. In the helper, I would use a combination of `getDepth()` and `getParent` to verify the `writePathRoot` is an ancestor of `writePath`.